tests/integration/metrics/column_pair_values/test_in_set.py

```diff
@@ -0,0 +1,149 @@
+COL_B_WITH_NULLS = "B_WITH_NULLS"
+DATA_FRAME = pd.DataFrame(
+    {
+        "id": [1, 2, 3, 4],
+        COL_A: [1, 2, 3, 4],
+        COL_B: ["a", "b", "c", "d"],
+    },
+)
+DATA_FRAME_WITH_NULLS = pd.DataFrame(
+    {
+        COL_A_WITH_NULLS: [None, 2, None, 4, None, None],
+        COL_B_WITH_NULLS: ["a", "b", None, None, None, None],
+    }
+)
+SUCCESS_VALUE_PAIR_SET = {(1, "a"), (2, "b"), (3, "c"), (4, "d")}
+FAILURE_VALUE_PAIR_SET = {(1, "a"), (2, "b"), (3, "c")}
+FAILURE_RESULT_COUNT = 1
+NO_MATCH_PAIR_SET = {(5, "e")}
+
+
+class TestColumnPairValuesInSetUnexpectedValues:
+    @parameterize_batch_for_data_sources(
+        data_source_configs=PANDAS_DATA_SOURCES + SPARK_DATA_SOURCES + SQL_DATA_SOURCES,
+        data=DATA_FRAME,
+    )
+    def test_success(self, batch_for_datasource: Batch) -> None:
+        batch = batch_for_datasource
+        metric = ColumnPairValuesInSetUnexpectedCount(
+            batch_id=batch.id,
+            value_pairs_set=SUCCESS_VALUE_PAIR_SET,
+            column_A=COL_A,
+            column_B=COL_B,
+        )
+        result = batch.compute_metrics(metric)
+        assert isinstance(result, ColumnPairValuesInSetUnexpectedCountResult)
+        assert result.value == 0
+
+    @parameterize_batch_for_data_sources(
+        data_source_configs=PANDAS_DATA_SOURCES + SPARK_DATA_SOURCES + SQL_DATA_SOURCES,
+        data=DATA_FRAME,
+    )
+    def test_failure(self, batch_for_datasource: Batch) -> None:
+        batch = batch_for_datasource
+        metric = ColumnPairValuesInSetUnexpectedCount(
+            batch_id=batch.id,
+            value_pairs_set=FAILURE_VALUE_PAIR_SET,
+            column_A=COL_A,
+            column_B=COL_B,
+        )
+        result = batch.compute_metrics(metric)
+        assert isinstance(result, ColumnPairValuesInSetUnexpectedCountResult)
+        assert result.value == FAILURE_RESULT_COUNT
+
+    @pytest.mark.parametrize(
+        "ignore_row_if,unexpected_count",
+        [
+            ("either_value_is_missing", 1),
+            ("both_values_are_missing", 3),
+            ("neither", 6),
+        ],
+    )
+    @parameterize_batch_for_data_sources(
+        data_source_configs=PANDAS_DATA_SOURCES,
+        data=DATA_FRAME_WITH_NULLS,
+    )
+    def test_ignore_row_if__pandas(
+        self, batch_for_datasource: Batch, ignore_row_if, unexpected_count
+    ) -> None:
+        batch = batch_for_datasource
+        metric = ColumnPairValuesInSetUnexpectedCount(
+            batch_id=batch.id,
+            value_pairs_set=NO_MATCH_PAIR_SET,
+            column_A=COL_A_WITH_NULLS,
+            column_B=COL_B_WITH_NULLS,
+            ignore_row_if=ignore_row_if,
+        )
+        result = batch.compute_metrics(metric)
+        assert result.value == unexpected_count
+
+    @pytest.mark.parametrize(
+        "ignore_row_if,unexpected_count",
+        [
+            ("either_value_is_missing", 2),
+            ("both_values_are_missing", 6),
```

Comment:
Are these 2 cases buggy? I would expect this to be the same as pandas. If so, could you change this to:
```
    @pytest.mark.parametrize(
        "ignore_row_if,unexpected_count",
        [
            pytest.param("either_value_is_missing", 1,  marks.pytest.mark.xfail),  # returns 2
            pytest.param("both_values_are_missing", 3, marks.pytest.mark.xfail),  # returns 6
            pytest.param("neither", 6),
        ],
    )
```
If this is caused by expected spark backend behavior, can you keep the test as is but add a comment you do for sql:
```
("either_value_is_missing", 2), # Spark ...
```

Reply:
sure - they seem to be buggy, but i'm not sure what's causing it. i'll add the xfails 👍

Reply:
I'm guessing Spark is inferring the wrong column type due to the `None`s, but I'm not 100% sure. Pretty sure I've encountered this in the past.